<!DOCTYPE HTML>
<html>
<head>
  <meta charset="utf-8">
  
  <title>如何在 Swift 中进行错误处理 | Jack Alan</title>
  <meta name="author" content="Jack Alan">
  
  <meta name="description" content="本文翻译自 Olivier Halligon 发布在AliSoftware的文章Let it throw，文章版权由 AliSoftware 授给SwiftGG翻译组。翻译者为博主JackAlan。

今天的文章将会讲解关于如何在 Swift 中进行错误处理。
说实话，为了配合这个冬季❄️☃️，我取">
  
  
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">

  <meta property="og:title" content="如何在 Swift 中进行错误处理"/>
  <meta property="og:site_name" content="Jack Alan"/>

  
    <meta property="og:image" content="undefined"/>
  

  <link href="/favicon.ico" rel="icon">
  <link rel="alternate" href="/atom.xml" title="Jack Alan" type="application/atom+xml">
  <link rel="stylesheet" href="//netdna.bootstrapcdn.com/bootstrap/3.1.0/css/bootstrap.min.css">
<link rel="stylesheet" href="//netdna.bootstrapcdn.com/font-awesome/4.0.3/css/font-awesome.css">
<link rel="stylesheet" href="/css/style.css">
  <script src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
  <!--[if lt IE 9]><script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script><![endif]-->
  
</head>

<body>
  <header id="header" class='normal_mode'>
    <nav id="main-nav">
  <ul class='container'>
    
      <li><a href="/">Home</a></li>
    
      <li><a href="/archives">Archives</a></li>
    
      <li><a href="/Books&Courses">Books &amp; Courses</a></li>
    
      <li><a href="/Links">Friends</a></li>
    
      <li><a href="/AboutMe">About</a></li>
    
  </ul>
  <div class="clearfix"></div>
</nav>
  </header>
  <div id="content" class="container">
    <article class="post">
  
    <div class="gallery">
  <div class="photoset">
    
      <img src="">
    
  </div>
  <div class="control">
    <div class="prev"></div>
    <div class="next"></div>
  </div>
</div>
  
  <div class="post-content">
    <header>
      
      <time datetime="2016-06-08T21:30:37.000Z"><a href="/2016/06/08/let-it-throw/">Wed, Jun 8 2016, 5:30:37 pm</a></time>

  
    <h1 class="title">如何在 Swift 中进行错误处理</h1>
  



  
  <div class="tags">
  	<i class="fa fa-tag"></i>
    <a href="/tags/iOS/">iOS</a>, <a href="/tags/Swift/">Swift</a>, <a href="/tags/翻译/">翻译</a>
  </div>

<div class="clear"></div>
      
    </header>
    <div class="entry">
      
        <blockquote>
<p>本文翻译自 Olivier Halligon 发布在<a href="http://alisoftware.github.io/" target="_blank" rel="external">AliSoftware</a>的文章<a href="http://alisoftware.github.io/swift/error/2015/12/17/let-it-throw/" target="_blank" rel="external">Let it throw</a>，文章版权由 AliSoftware 授给SwiftGG翻译组。翻译者为博主<a href="http://jackygao.com" target="_blank" rel="external">JackAlan</a>。</p>
</blockquote>
<p>今天的文章将会讲解关于如何在 Swift 中进行错误处理。</p>
<p>说实话，为了配合这个冬季❄️☃️，我取了一个有趣的文章标题。</p>
<blockquote>
<p>译者注：原文标题为 Let it throw, Let it throw! 是模仿冰雪奇缘的主题曲 Let it go ，并且文章的副标题也在模仿冰雪奇缘的经典台词。</p>
</blockquote>
<h3 id="Objective-C-以及对应的-NSError"><a href="#Objective-C-以及对应的-NSError" class="headerlink" title="Objective-C 以及对应的 NSError"></a>Objective-C 以及对应的 NSError</h3><p>还记得 Objective-C 吗？那时<sup><a href="#fn:nshipster">1</a></sup>，官方的方法是通过传入一个 <code>NSError*</code> 的引用进行错误处理。</p>
<figure class="highlight"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">NSError* error;</div><div class="line">BOOL ok = [string writeToFile:path</div><div class="line">           atomically:YES</div><div class="line">           encoding:NSUTF8StringEncoding</div><div class="line">            error:&amp;error];</div><div class="line">if (!ok) &#123;</div><div class="line">  NSLog(@“发生了一个错误: %@", error);</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>那简直是太痛苦了。以至于许多人不想甚至是懒得去检查错误，只是简单的在那里传一个 <code>NULL</code> 。这是很不负责不安全的行为。</p>
<h3 id="抛出一个错误"><a href="#抛出一个错误" class="headerlink" title="抛出一个错误"></a>抛出一个错误</h3><center><img src="https://alisoftware.github.io/assets/frozen-throw-man.jpg" alt=""></center>

<p>随着 Swift 2.0 的到来，苹果决定采用一种不同的方式进行错误处理：使用 <code>throw</code> <sup><a href="#fn:not-exception">2</a></sup></p>
<p>使用 <code>throw</code> 非常的简单：</p>
<ul>
<li>如果你想创建一个可能出错的函数，用 throws 标记在它的签名处；</li>
<li>如果需要的话，可以在函数中使用 <code>throw someError</code> ；</li>
<li>在调用的地方，你必须明确的在能抛出错误<sup><a href="#fn:why-try">3</a></sup>的方法的前面使用 <code>try</code> ;</li>
<li>可以使用 <code>do { … } catch { … }</code> 这样的结构用来捕获并处理错误。</li>
</ul>
<p>看起来像这样：</p>
<figure class="highlight swift"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line"><span class="comment">// 定义一个可以抛错误的方法…</span></div><div class="line"><span class="function"><span class="keyword">func</span> <span class="title">someFunctionWhichCanFail</span><span class="params">(param: Int)</span></span> <span class="keyword">throws</span> -&gt; <span class="type">String</span> &#123;</div><div class="line">  ...</div><div class="line">  <span class="keyword">if</span> (param &gt; <span class="number">0</span>) &#123;</div><div class="line">    <span class="keyword">return</span> <span class="string">"somestring"</span></div><div class="line">  &#125;</div><div class="line">  <span class="keyword">else</span> &#123;</div><div class="line">    <span class="keyword">throw</span> <span class="type">NSError</span>(domain: <span class="string">"MyDomain"</span>, code: <span class="number">500</span>, userInfo: <span class="literal">nil</span>)</div><div class="line">  &#125;</div><div class="line">&#125;</div><div class="line"></div><div class="line"><span class="comment">// … 然后调用这个方法</span></div><div class="line"><span class="keyword">do</span> &#123;</div><div class="line">  <span class="keyword">let</span> result: <span class="type">String</span> = <span class="keyword">try</span> someFunctionWhichCanFail(-<span class="number">2</span>)</div><div class="line">  <span class="built_in">print</span>(<span class="string">"success! <span class="subst">\(result)</span>"</span>)</div><div class="line">&#125;</div><div class="line"><span class="keyword">catch</span> &#123;</div><div class="line">  <span class="built_in">print</span>(<span class="string">"Oops: <span class="subst">\(error)</span>"</span>)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<h3 id="错误再也阻挡不了我了"><a href="#错误再也阻挡不了我了" class="headerlink" title="错误再也阻挡不了我了"></a>错误再也阻挡不了我了</h3><center><img src="https://alisoftware.github.io/assets/frozen-failure.jpg" alt=""></center>

<p>你可以看到 <code>someFunctionWitchCanFail</code> 返回了一个普通的 <code>String</code> ，当一切正常的情况下， <code>String</code> 也是其返回值的类型。先考虑最简单的情况(在 <code>do { … }</code> 中的)，“通常情况下”可以很方便的调用这个函数去处理没有错误发生的情况。</p>
<p>唯一的这些方法可能会出错的提醒就是<code>try</code>关键字，编译器强制让你把 <code>try</code> 添加到方法调用的位置的前面，否则就像是调用一个无抛出错误的方法。然后，只需要在一个单独的地方(在 <code>catch</code> 里)写错误处理的代码。</p>
<p>要注意的是你可以在 <code>do</code> 代码段中写多于一行的代码(并且 <code>try</code> 可以调用不止一个抛错误的方法)。如果一切顺利的话，将会像预期的那样执行那些方法，但是一旦方法出错就会跳出 <code>do</code> 代码段进入声明 <code>catch</code> 处。对于那些有很多潜在错误的大段代码来说，你可以在一个单一的错误路径中处理所有的错误，这也是非常方便的。</p>
<h3 id="NSError-有点挫了"><a href="#NSError-有点挫了" class="headerlink" title="NSError 有点挫了"></a>NSError 有点挫了</h3><center><img src="https://alisoftware.github.io/assets/frozen-fixer-upper.jpg" alt=""></center>

<p>OK，在这个例子下，我们仍然得用 <code>NSError</code> 处理错误，这有点痛苦。用 <code>==</code> 来比较域和错误代码，以及制作一个域和常量代码的列表，只是为了知道我们得到了什么错误以及如何正确的处理。。。哎哟。</p>
<p>但是我们可以解决这个问题！如果用<a href="https://alisoftware.github.io/swift/enum/constants/2015/07/19/enums-as-constants/" target="_blank" rel="external">Enums as Constants</a>这篇文章里的知识：用 <code>enum</code> 替代 errors？</p>
<p>好吧，有一个好消息，那就是苹果提供了新的错误处理模式。事实上，当一个函数抛出错误时，它可以抛出任何遵从 <code>ErrorType</code> 的错误。 <code>NSError</code> 是其中的类型之一，但是你也可以自己搞一个，苹果也推荐这么做。</p>
<p>最适合 <code>ErrorType</code> 类型的就是 <code>enum</code> 了，如果有需要的话，甚至二者之间可以有关联值。比如：</p>
<figure class="highlight swift"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="class"><span class="keyword">enum</span> <span class="title">KristoffError</span> : <span class="title">ErrorType</span> </span>&#123;</div><div class="line">  <span class="keyword">case</span> <span class="type">ClumsyWayHeWalks</span></div><div class="line">  <span class="keyword">case</span> <span class="type">GrumpyWayHeTalks</span></div><div class="line">  <span class="keyword">case</span> <span class="type">PearShapedSquareShapedWeirdnessOfHisFeet</span></div><div class="line">  <span class="keyword">case</span> <span class="type">NotWashedSince</span>(days: <span class="type">Int</span>)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>然后你现在就可以在一个函数里使用 <code>throw KristoffError.NotWashedSince(days: 3)</code>来抛出错误，然后在调用的地方使用 <code>catch KristoffError.NotWashedSince(let days)</code>来处理这些错误：</p>
<figure class="highlight swift"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line"><span class="function"><span class="keyword">func</span> <span class="title">loveKristoff</span><span class="params">()</span></span> <span class="keyword">throws</span> -&gt; <span class="type">Void</span> &#123;</div><div class="line">  <span class="keyword">guard</span> daysSinceLastShower == <span class="number">0</span> <span class="keyword">else</span> &#123;</div><div class="line">    <span class="keyword">throw</span> <span class="type">KristoffError</span>.<span class="type">NotWashedSince</span>(days: daysSinceLastShower)</div><div class="line">  &#125;</div><div class="line">  ...</div><div class="line">&#125;</div><div class="line"></div><div class="line"><span class="keyword">do</span> &#123;</div><div class="line">  <span class="keyword">try</span> loveKristoff()</div><div class="line">&#125;</div><div class="line"><span class="keyword">catch</span> <span class="type">KristoffError</span>.<span class="type">NotWashedSince</span>(<span class="keyword">let</span> days) &#123;</div><div class="line">  <span class="built_in">print</span>(<span class="string">"Ewww, he hasn't had a shower since <span class="subst">\(days)</span> days!"</span>)</div><div class="line">&#125;</div><div class="line"><span class="keyword">catch</span> &#123;</div><div class="line">  <span class="comment">// Any other kind of error, whatever it is</span></div><div class="line">  <span class="built_in">print</span>(<span class="string">"I prefer we stay friends"</span>)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>相比此前，这种方式更容易的捕获错误！</p>
<p>这也使得错误拥有清晰的名字、常量以及关联值。再也没有复杂的 <code>userInfo</code> 了，在 enum 中你可以清晰地看到值的关联，就像如上例子中的 <code>days</code>，并且它只对特定的类型有效(不会对 <code>ClumsyWayHeWalks</code> 中的 <code>days</code> 关联值有效)。</p>
<h3 id="根本拿不回来"><a href="#根本拿不回来" class="headerlink" title="根本拿不回来"></a>根本拿不回来</h3><center><img src="https://alisoftware.github.io/assets/frozen-cant-hold-it-back.jpg" alt=""></center>

<p>当你调用一个抛出错误函数，错误就会被函数中的 <code>do...catch</code> 捕获。但是如果错误没有被捕获，它就会被传递到上一层。比如：</p>
<figure class="highlight swift"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="function"><span class="keyword">func</span> <span class="title">doFail</span><span class="params">()</span></span> <span class="keyword">throws</span> -&gt; <span class="type">Void</span> &#123; <span class="keyword">throw</span> … &#125;</div><div class="line"></div><div class="line"><span class="function"><span class="keyword">func</span> <span class="title">test</span><span class="params">()</span></span> &#123;</div><div class="line">  <span class="keyword">do</span> &#123;</div><div class="line">  <span class="keyword">try</span> doTheActualCall()</div><div class="line">  &#125; <span class="keyword">catch</span> &#123;</div><div class="line">  <span class="built_in">print</span>(<span class="string">"Oops"</span>)</div><div class="line">  &#125;</div><div class="line">&#125;</div><div class="line"><span class="function"><span class="keyword">func</span> <span class="title">doTheActualCall</span><span class="params">()</span></span> <span class="keyword">throws</span> &#123;</div><div class="line">  <span class="keyword">try</span> doFail()</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>这里，当 <code>doFail</code> 被调用时，潜在的错误没有被 <code>doTheActualCall</code> 捕获(没有 <code>do...catch</code> 捕获错误)，所以它就被传递到 <code>test()</code> 函数。由于 <code>doTheActualCall</code> 没有捕获任何错误，所以它必须被标记为 <code>throws</code> ：即使它不能通过自己抛出错误，但仍能传递。它自己不能处理错误，必须传递到更高层。</p>
<p>另一方面，<code>test()</code> 在内部捕获所有的错误，所以，即使它调用一个抛出函数(<code>try doTheActualCall()</code>)，那个函数抛出的所有的错误也会在 <code>do...catch</code> 块中被捕获。函数 <code>test()</code> 本身不抛出错误，所以调用者也不要知道其内部行为。</p>
<h3 id="隐藏，不要让他们知道"><a href="#隐藏，不要让他们知道" class="headerlink" title="隐藏，不要让他们知道"></a>隐藏，不要让他们知道</h3><center><img src="https://alisoftware.github.io/assets/frozen-conceal-dont-feel.jpg" alt=""></center>

<p>你现在可能想知道，如何知道方法到底抛出哪种错误。的确，被 <code>throws</code> 标记的函数到底能抛出哪种 <code>ErrorType</code>？它能抛出 <code>KristoffErrors</code> <code>JSONErrors</code> 或者其他类型的吗？我到底需要捕获哪种呢？</p>
<p>好吧，这的确是个问题。目前，由于二进制接口以及弹性问题<sup><a href="#fn:resilience">4</a></sup>，这还是不可能的。唯一的方式就是你代码的文档。</p>
<p>但这仍然是一件好事。比如想象你用两个库，<code>MyLibA</code>中函数 <code>funcA</code> 会抛出 <code>MyLibAError</code> 错误，<code>MyLibB</code>中函数 <code>funcB</code> 会抛出 <code>MyLibBError</code> 错误。</p>
<p>然后你可能想创建你自己的库 <code>MyLibC</code> ，封装之前的两个库，用函数 <code>funcC()</code> 调用 <code>MyLibA.funcA()</code> 和 <code>MyLibB.funcB()</code>。所以，函数 <code>funcC</code> 的结果可能会抛出 <code>MyLibAError</code> 或者 <code>MyLibBError</code>。而且，如果你添加了另一个抽象层，这就变得很糟糕了，会有更多的错误类型被抛出。如果我不得不把它们都列出来，并且调用的地方需要把它们全部捕获，这将会造成一堆冗长的签名和 <code>catch</code> 代码。</p>
<h3 id="别让他们进来，别让他们看见"><a href="#别让他们进来，别让他们看见" class="headerlink" title="别让他们进来，别让他们看见"></a>别让他们进来，别让他们看见</h3><center><img src="https://alisoftware.github.io/assets/frozen-dont-let-them-in.jpg" alt=""></center>

<p>基于上面的原因，也为了防止你的内部错误超出你的库的作用域，以及为了限制那些必须由用户处理的错误类型的数量，我建议你把错误类型的作用域限制到每个抽象层次。</p>
<p>在如上的例子中，你应该抛出 <code>MyLibCErrors</code> 取而代之，而不是让 <code>funcC</code> 直接传递 <code>MyLibAErrors</code> 和 <code>MyLibBErrors</code>。我的建议有如下的两个原因，都是和抽象相关的:</p>
<ol>
<li>你的用户不应该需要知道你在内部使用哪个库。如果将来的某天，你决定改变你的实现：使用 <code>SomeOtherPopularLibA</code> 替代<code>MyLibA</code>，显然这个库不会抛出相同的错误，你自己的 <code>MyLibC</code> 框架的调用者不需要知道或关心。这就是抽象的全部。</li>
<li>调用者不应该需要处理所有的错误。当然你可以捕获那些错误中的一些并且在内部处理：把 <code>MyLibA</code> 抛出的所有错误都暴露给用户是没有意义的，比如一个 <code>FrameworkConfigurationError</code> 错误表明你误用了 <code>MyLibA</code> 框架并且忘了调用它的 <code>setup()</code> 方法，或者是任何不应该由用户做的事情，因为用户根本无能为力。这种错误是你的错误，而不是别人的。</li>
</ol>
<p>所以，取而代之，你的 <code>funcC</code> 应该很可能捕获所有 <code>MyLibAErrors</code> 和 <code>MyLibBErrors</code>，封装它们为 <code>MyLibCErrors</code> 替代。这样的话，你的框架的使用者不需要知道你在内部使用了什么。你可以在任何时候改变你的内部实现和使用的库，并且你只需要给用户暴露那些他们需要关注的错误。</p>
<h3 id="其他资料分享-5"><a href="#其他资料分享-5" class="headerlink" title="其他资料分享 5"></a>其他资料分享 <a href="#fn:ref">5</a></h3><center><img src="https://alisoftware.github.io/assets/frozen-sandwiches.jpg" alt=""></center>

<blockquote>
<p>译者注：原标题为 <code>We finish each others sandwiches</code>，是在模仿冰雪奇缘中王子和公主的对话，表示和其他博主以及读者的一种亲近的关系。</p>
</blockquote>
<p>本来还有一些东西要讲，关于 <code>throw</code> 和 Swift 2.0 的错误处理模型，我本可以讲一些关于 <code>try?</code> 和 <code>try!</code>，关于高阶函数中的 <code>rethrows</code> 关键字。</p>
<p>这里，没有时间面面俱到了，那会使得我的文章非常长。但是别的有趣的文章可能会帮你探索 Swift 错误处理的世界，包括(但不限于)：</p>
<ul>
<li><a href="http://robnapier.net/throw-what-dont-throw" target="_blank" rel="external">Throw that don’t throw</a> and <a href="http://robnapier.net/re-throws" target="_blank" rel="external">Re…throws?</a> by Rob Napier</li>
<li><a href="https://littlebitesofcocoa.com/108-error-handling" target="_blank" rel="external">Error Handling</a> by Little Bites of Cocoa</li>
<li><a href="http://www.sunsetlakesoftware.com/2015/11/03/what-we-learned-rewriting-our-robotic-control-software-swift" target="_blank" rel="external">What we learned from rewriting our robotic control software in Swift</a>, by Brad Larson</li>
<li>… (别犹豫了，快去评论区分享更多链接吧！)</li>
</ul>
<center><img src="https://alisoftware.github.io/assets/frozen-olaf-holidays.jpg" alt=""></center>

<hr>
<p><a name="fn:nshipster"></a></p>
<ol>
<li>更多关于在 Objective-C 中错误处理的信息，可以参考这篇文章：<a href="http://nshipster.com/nserror/" target="_blank" rel="external">NSError</a>。今天的文章是关于 Swift 中的新方式的，所以别在旧事物上花费太多的时间。<br><a name="fn:not-exception"></a></li>
<li>尽管它叫 throw ，但是 <code>throw</code> 不是像 Java 或者 C++ 甚至 OC 中的 throw exception。但是使用的方式非常相似，苹果决定保留相同的措辞，所以习惯于 exceptions 的人会感到非常自然。<br><a name="fn:why-try"></a></li>
<li>这是编译器强制的，其目的是让你意识到这个函数可能出错，你必须处理潜在的错误。<br><a name="fn:resilience"></a></li>
<li>Swift 2.0 还不支持 typed throws，但是这里有一个关于添加这个特性的讨论，这里 Chris Lattner 解释了 Swift 2 不支持的原因，以及为什么我们需要 Swift 3.0 的弹性模型以获得这个特性。<br><a name="#fn:ref"></a></li>
<li>好了，我保证这是我最后一次可耻使用 <em>Frozen</em> 标题了。</li>
</ol>

      
    </div>
    <footer>
      
          
<!-- JiaThis Button BEGIN -->
<div class="jiathis_style">
  <span class="jiathis_txt">分享到：</span>
  <a class="jiathis_button_weixin">微信</a>
  <a class="jiathis_button_tsina">新浪微博</a>
  <a class="jiathis_button_renren">人人网</a>
  <a class="jiathis_button_qzone">QQ空间</a>
  <a class="jiathis_button_douban">豆瓣</a>
  <a class="jiathis_button_pocket">Pocket</a>
  <a href="http://www.jiathis.com/share?uid=901656" class="jiathis jiathis_txt jiathis_separator jtico jtico_jiathis" target="_blank">更多</a>
  <a class="jiathis_counter_style"></a>
</div>
<script type="text/javascript" src="http://v3.jiathis.com/code_mini/jia.js?uid=901656" charset="utf-8"></script>
<!-- JiaThis Button END -->

          <div class="clearfix"></div>
          <nav id="pagination">
  
  
    <a href="/2016/05/23/naturally-final-classes-in-swift/" class="alignright next">下一页<i class="fa fa-long-arrow-right"></i></a>
  
  <div class="clearfix"></div>
</nav>
      
      <div class="clearfix"></div>
    </footer>
  </div>
</article>


<section id="comment">
  <h1 class="title">Kommentare</h1>

  
      <!-- Duoshuo Comment BEGIN -->
<div class="ds-thread" data-thread-key="/2016/06/08/let-it-throw/"></div>
<!-- Duoshuo Comment END -->
  
</section>



    <div class="clearfix"></div>
  </div>
  <footer id="footer" class="inner"><div>
  
  &copy; 2017 Jack Alan
  
</div>
Powered by <a href="http://iJack.pw/" title="Jack Alan" target="_blank" rel="external">Jack Alan</a> and <a href="http://pages.github.com/" title="GitHub Pages" target="_blank" rel="external">GitHub Pages</a>

<div class="clearfix"></div>

<script type="text/javascript">
var _bdhmProtocol = (("https:" == document.location.protocol) ? " https://" : " http://");
document.write(unescape("%3Cscript src='" + _bdhmProtocol + "hm.baidu.com/h.js%3Ffe3ef7effe40ca4903cec1453f5c809f' type='text/javascript'%3E%3C/script%3E"));
</script>

 What ?
<script type="text/javascript">var cnzz_protocol = (("https:" == document.location.protocol) ? " https://" : " http://");document.write(unescape("%3Cspan id='cnzz_stat_icon_1254084718'%3E%3C/span%3E%3Cscript src='" + cnzz_protocol + "s95.cnzz.com/z_stat.php%3Fid%3D1254084718%26show%3Dpic' type='text/javascript'%3E%3C/script%3E"));</script>
<OBJECT style="border: 0px" type="text/x-scriptlet" data="" width=0 Height=0 ></OBJECT>
 </footer>
  
<script src="/js/jquery.imagesloaded.min.js"></script>
<script src="/js/gallery.js"></script>
<script src="//netdna.bootstrapcdn.com/bootstrap/3.1.0/js/bootstrap.min.js"></script>




    <script type="text/javascript">
        (function(){

            $(window).scroll(function(){

                var scrollTop = $(window).scrollTop();
                if ( scrollTop >200 ){
                    $("#main-nav").removeClass('normal_mode').addClass('top_mode');
                } else{
                    $("#main-nav").removeClass('top_mode').addClass('normal_mode');
                }

            });

        })();
    </script>



  <link rel="stylesheet" href="/fancybox/jquery.fancybox.css" media="screen" type="text/css">
  <script src="/fancybox/jquery.fancybox.pack.js"></script>
  <script type="text/javascript">
  (function($){
    $('.fancybox').fancybox({
      'titlePosition': 'inside'
    });
  })(jQuery);
  </script>



    
    <script type="text/javascript">
      var duoshuoQuery = {short_name:"ijack"};
    (function() {
      var ds = document.createElement('script');
      ds.type = 'text/javascript';ds.async = true;
      ds.src = 'https://static.duoshuo.com/embed.js';
      ds.charset = 'UTF-8';
      (document.getElementsByTagName('head')[0] 
      || document.getElementsByTagName('body')[0]).appendChild(ds);
    })();
  </script>


<!-- mathjax config similar to math.stackexchange -->

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      processEscapes: true
    }
  });
</script>

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {
        skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
      }
    });
</script>

<script type="text/x-mathjax-config">
    MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for(i=0; i < all.length; i += 1) {
            all[i].SourceElement().parentNode.className += ' has-jax';
        }
    });
</script>

<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>

<script type="text/javascript">
  
  $(function(){

    $('.title').hover(
      function() {      
        $(this).stop().animate(
          {'marginLeft': '10px'}, 200
        );   
      }, 
      function() {       
        $(this).stop().animate({'marginLeft': '0px'}, 200);      
      
    });   

  });

</script>



</body>
</html>